arrow-core-data/src/test/kotlin/arrow/core/EitherTest.kt

```diff
@@ -203,5 +210,83 @@ class EitherTest : UnitSpec() {
           Right(a).handleErrorWith { Right(b) } == Right(a)
       }
     }
+
+    "suspended Either.fx can bind immediate values" {
+      Gen.either(Gen.string(), Gen.int())
+        .random()
+        .take(1001)
+        .forEach { either ->
+          Either.fx<String, Int> {
+            val res = !either
+            res
+          } shouldBe either
+        }
+    }
+
+    "suspended Either.fx can bind suspended values" {
+      Gen.either(Gen.string(), Gen.int())
+        .random()
+        .take(1001)
+        .forEach { either ->
+          Either.fx<String, Int> {
+            val res = !(suspend {
+              sleep(1)
+              either
+            }).invoke()
+
+            res
+          } shouldBe either
+        }
+    }
+
+    "suspended Either.fx can safely handle immediate exceptions" {
+      Gen.bind(Gen.int(), Gen.throwable(), ::Pair)
+        .random()
+        .take(1001)
+        .forEach { (i, exception) ->
+          shouldThrow<Throwable> {
+            Either.fx<String, Int> {
+              val res = !Either.Right(i)
+              throw exception
+              res
```

Comment:
We could consider testing a case where we perform a side effect here that should never happen to verify that the execution is actually stopped, what do you think?